core/state/trie_prefetcher.go

```diff
@@ -0,0 +1,396 @@
+				}
+				p.taskLock.Unlock()
+			}
+		}
+	}
+}
+
+// Close stops the prefetcher
+func (p *TriePrefetcher) Close() {
+	if p.quitCh != nil {
+		close(p.quitCh)
+		p.quitCh = nil
+	}
+}
+
+// Resume causes the prefetcher to clear out old data, and get ready to
+// fetch data concerning the new root.
+func (p *TriePrefetcher) Resume(root common.Hash) {
+	// Abort if the prefetcher is not paused
+	if atomic.LoadUint32(&p.paused) == 0 {
+		log.Error("Trie prefetcher already resumed")
+		return
+	}
+	atomic.StoreUint32(&p.paused, 0)
+
+	cmd := &command{
+		root: root,
+		done: make(chan struct{}),
+	}
+	p.cmdCh <- cmd
+	<-cmd.done
+}
+
+// Pause causes the prefetcher to pause prefetching, and make tries
+// accessible to callers via GetTrie.
+func (p *TriePrefetcher) Pause() {
+	// Abort if the prefetcher is already paused
+	if atomic.LoadUint32(&p.paused) == 1 {
+		log.Error("Trie prefetcher already paused")
+		return
+	}
+	atomic.StoreUint32(&p.paused, 1)
+
+	// Request a pause and wait until it's done
+	cmd := &command{
+		done: make(chan struct{}),
+	}
+	p.cmdCh <- cmd
+	<-cmd.done
+}
+
+// PrefetchAddresses adds an address for prefetching
+func (p *TriePrefetcher) PrefetchAddresses(addrs []common.Address) {
+	// Abort if the prefetcher is already paused
+	if atomic.LoadUint32(&p.paused) == 1 {
+		log.Error("Attempted account trie-prefetch whilst paused")
+		return
+	}
+	// Inject the addresses into the task queue and notify the prefetcher
+	p.taskLock.Lock()
+	defer p.taskLock.Unlock()
+
+	p.accountTasks = append(p.accountTasks, addrs...)
+	select {
+	case p.reqCh <- struct{}{}:
+	default:
+		// Already notified
+	}
+}
+
+// PrefetchStorage adds a storage root and a set of keys for prefetching
+func (p *TriePrefetcher) PrefetchStorage(root common.Hash, slots []common.Hash) {
+	// Abort if the prefetcher is already paused
+	if atomic.LoadUint32(&p.paused) == 1 {
+		log.Error("Attempted storage trie-prefetch whilst paused")
+		return
+	}
+	// Inject the storage hashes into the task queue and notify the prefetcher
+	p.taskLock.Lock()
+	defer p.taskLock.Unlock()
+
+	p.storageTasks[root] = append(p.storageTasks[root], slots...)
+	select {
+	case p.reqCh <- struct{}{}:
+	default:
+		// Already notified
+	}
+}
+
+// GetTrie returns the trie matching the root hash, or nil if the prefetcher
+// doesn't have it.
+func (p *TriePrefetcher) GetTrie(root common.Hash) Trie {
+	// Abort if the prefetcher is not paused
+	if atomic.LoadUint32(&p.paused) == 0 {
+		log.Error("Attempted trie-prefetcher retrieval whilst not paused")
+		return nil
+	}
+	if root == p.accountTrieRoot {
+		return p.accountTrie
+	}
+	if storageTrie, ok := p.storageTries[root]; ok {
+		// Two accounts may well have the same storage root, but we cannot allow
+		// them both to make updates to the same trie instance. Therefore,
+		// we need to either delete the trie now, or deliver a copy of the trie.
+		delete(p.storageTries, root)
+		return storageTrie
+	}
```

Comment:
@karalabe I've been thinking about this. The current `Trie` implementation already copies on modifications, so if we just copy the trie root node before delivering it, I think it would be fine if two accounts have the same storage root and obtain the "same" trie. 
If we were to use a less copy-happy trie implementation, we'd have to do the deletion here, but we're not.

Reply:
True, though I don't expect to see a gain (i.e. the chance of two tries being the same is more of a deliberate attack than a thing that would normally happen). Still, the code would get cleaner.